### PR TITLE
FixedLayout

### PR DESCRIPTION
**Pull request by rylev**
This creates a FixedLayout marker trait for any type that has a fixed layout.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1,7 +1,15 @@
...
+/// representation for Rust types
+pub unsafe trait FixedLayout {}
+pub fn ensure_fixed_layout<T: FixedLayout>() {}
+
+macro_rules! fixed_layout_impl {
+    ($($type:ty),*) => {
+        $(unsafe impl FixedLayout for $type {})*
+    };
 }
+fixed_layout_impl!(u8, u16, u32, u64, u128, i8, i16, i32, i64, i128, bool);
```
Comment (AndrewGaspar) [inline src/lib.rs:15]: Here are some other types that seem like they should be included (maybe you're not for now on purpose?):
* `usize` and `isize` (or does `FixedLayout` require no platform differences?)
* `NonZero*`
* `Option<NonZero*>`

Comment (AndrewGaspar) [reply src/lib.rs:15]: Also, should `FixedLayout` apply to raw pointer types (`*const i32`)?

Comment (RustyYato) [reply src/lib.rs:15]: Yes, raw pointers have a well defined layout, as do `references` and `Option<reference>`

Comment (rylev) [reply src/lib.rs:15]: Yes I'm not done. Happy to accept PRs 😄

Comment (AndrewGaspar) [reply src/lib.rs:15]: Gotcha, that's alright, just wanted to check my understanding.

Comment (rylev) [reply src/lib.rs:15]: @KrishnaSannasi I don't think that's entirely true. `&str` and `&[T]` aren't a fixed layout I believe...

Comment (RustyYato) [reply src/lib.rs:15]: Right, I forgot about `!Sized`, references do have a fixed layout when the pointee is `Sized`

Comment (rylev) [reply src/lib.rs:15]: Cool. I can add all these unless @AndrewGaspar wants to

Comment (AndrewGaspar) [reply src/lib.rs:15]: No, sorry, not right now, but maybe I'll contribute later. I'd like to use this crate as a public dependency of [another crate](https://github.com/rsmpi/rsmpi) once the API is semi-stable, so I'm just trying to keep an eye on the work here.

**Pull request merged**